### ISSUE TITLE
Limit scope of api keys to reduce account takeover risks

### ISSUE THREAD

**Issue by SatsAllDay** [stackernews/stacker.news#984]
**Is your feature request related to a problem? Please describe.**
With API Keys being introduced, I thought about the potential ramifications if an API Key got into the wrong hands (not from a SN leak, but if the user accidentally exposed it).

Obviously there is always some risk, such as draining the SN wallet funds, etc. But maybe there's some APIs that we can restrict access to via API keys?

**Describe the solution you'd like**
I think it might make sense to disallow certain API actions to be performed by API keys, most notably:
1. `unlinkAuth` - should an API Key be compromised, the attacker could use this API to remove existing auth methods, locking the account owner out of their account
2. Whatever APIs are used to register new auth methods, which when used with (1) above would allow a complete account takeover (I think)?
3. Since we're limited to a single API key right now, I don't think there's any reason to limit access to the `generateApiKey` and `deleteApiKey` APIs, since the attacker would basically be locking themself out if they deleted the only API key.

**Describe alternatives you've considered**
In the long term, following a model like OAuth scoping is probably a good idea, but I imagine we should not roll our own, or if we do, it is a much larger effort to:
1. Enumerate different scopes/roles
2. Assign APIs to each scope/role
3. Generically/abstractly enforce API Key scopes on each API 

This is more of a stop-gap/band-aid to give a little more security for us clumsy stackers ;) 

**Additional context**
N/A

As always, I'm open to input. Maybe I'm making a bigger deal out of this than is warranted.

**Comment by ekzyis** [comment]
> As always, I'm open to input. Maybe I'm making a bigger deal out of this than is warranted.

No, I think it's great that you wrote this down. I agree, we should do this at some point but I don't think it's a priority right now. We could add some warnings how sensitive this API key is in the short term though.

Regarding implementation: I have some experience with [`@casl/ability`](https://casl.js.org/v6/en/guide/intro) and I think it was easy to use so I think we would make no mistake with trying to use it for this. [It even integrates with Prisma via `@casl/prisma`](https://casl.js.org/v6/en/package/casl-prisma) though I am not sure how long we'll continue using Prisma as part of our queries.

[It's also what NestJS recommends to use](https://docs.nestjs.com/security/authorization). Meanwhile, the [authorization section of NextJS](https://nextjs.org/docs/pages/building-your-application/authentication#authorization) only mentions middleware which seem to be implemented as an afterthought ime. Middleware wouldn't work anyway for this in our case since it executes before we initialize sessions iirc.

In general, I think our existing "authorization layer" which consists of `if (!me)` in every resolver needs some love first before we can make it more complex. There is also a section about authorization on graphql.org [here](https://graphql.org/learn/authorization/).

**Comment by huumn** [comment]
To start, we could just outright prevent zapping/withdrawals and unlinking/linking of auth methods via an api key.

**Comment by SatsAllDay** [comment]
> To start, we could just outright prevent zapping/withdrawals and unlinking/linking of auth methods via an api key.

Makes sense! I'll self-assign this one :)
